Feature: save lock control via actions

adamsilverstein opened the pull request:
## Description
Add a post save locking feature. Plugins can add and remove locks and when any locks are present the publish and update buttons are disabled.

Supersedes https://github.com/WordPress/gutenberg/pull/10115
Fixes https://github.com/WordPress/gutenberg/issues/7020

## How has this been tested?
Load gutenberg and make some edits. In the console, type:
`wp.data.dispatch( 'core/editor' ).lockPostSaving( 'mylock' ); `
Note: the publish or update button becomes disabled.
Remove the post lock with:
`wp.data.dispatch( 'core/editor' ).unlockPostSaving( 'mylock' );`
* Test adding multiple locks: the button will remain disabled until all locks are removed.

## Types of changes
* Add actions for lock and unlock post saving.
* Add an isPostSavingLocked selector
* Check isPostSavingLocked in post publish button

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

adamsilverstein committed:
Add new actions for lockPostSaving and unlockPostSaving

adamsilverstein committed:
isPostSavingLocked selector

adamsilverstein committed:
postSavingLock reducer

adamsilverstein committed:
check isPostSavingLocked in post publish button

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -913,6 +913,38 @@ export function postLock( state = { isLocked: false }, action ) {
+export function postSavingLock( state = [], action ) {
+	const { lockName } = action;
+	switch ( action.type ) {
+		case 'LOCK_POST_SAVING':
```
I don't have a strong preference, but tracking this as an object could simplify the reducer implementation a fair bit:

```js
case 'LOCK_POST_SAVING':
case 'UNLOCK_POST_SAVING':
	return {
		...state,
		[ lockName ]: action.type === 'LOCK_POST_SAVING'
	};
```

adamsilverstein replied:
ah, good suggestion, this would simplify the structure, i'll give that a try

adamsilverstein replied:
resolved in https://github.com/WordPress/gutenberg/pull/10649/commits/8fea9a0b1665ee73bec67bad457bfa5c4cc9f8d4

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -1966,6 +1966,17 @@ export function isPostLocked( state ) {
+ * @return {boolean} Is locked.
+ */
+export function isPostSavingLocked( state ) {
+	return state.postSavingLock && state.postSavingLock.length > 0;
```
When will `state.postSavingLock` be falsey?

adamsilverstein replied:
possibly before it is set? This is me writing defensive programs. I can remove this check.

adamsilverstein replied:
Updated in https://github.com/WordPress/gutenberg/pull/10649/commits/9117ace44cf980cc3d10583e53f574ddad56d55d

aduth reviewed (changes requested):
Do we need to consider the post lock for autosaving?

We should have unit tests for at least selectors and reducer.

You need to run and commit changed files from `npm run docs:build`.

adamsilverstein commented:
> Do we need to consider the post lock for autosaving?

I don't think so, these should fire as usual. Autosaves protect the user from lost content.

> We should have unit tests for at least selectors and reducer.

I can add these, wanted to make sure approach was acceptable first.

> You need to run and commit changed files from npm run docs:build.

Ah, ok will do.

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into feature/save-lock-control

adamsilverstein committed:
simplify reducer with object

adamsilverstein committed:
adjust isPostSavingLocked for new state shape

adamsilverstein committed:
build docs

adamsilverstein commented:
@aduth I have addressed your feedback and this is ready for review.

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -913,6 +913,29 @@ export function postLock( state = { isLocked: false }, action ) {
+ *
+ * @return {PostLockState} Updated state.
+ */
+export function postSavingLock( state = [], action ) {
```
You changed the return shape for the lock actions, but the default is still an array.

adamsilverstein replied:
good point! addressed in https://github.com/WordPress/gutenberg/pull/10649/commits/e97e01a815fd3c50c4aebf65ae1e442a00948daf

adamsilverstein commented on packages/editor/src/store/reducer.js:
```diff
@@ -1133,4 +1156,5 @@ export default optimist( combineReducers( {
 	autosave,
 	settings,
 	tokens,
+	postSavingLock,
```
good idea! i was going to move it to fix the order then found there is no existing order i could discern.

adamsilverstein committed:
postSavingLock - initial state is an empty object

adamsilverstein commented:
I'll work on some tests.

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into feature/save-lock-control

adamsilverstein committed:
ensure the isPostSavingLocked selector returns a boolean

adamsilverstein committed:
Add tests for isPostSavingLocked

adamsilverstein committed:
Add tests for the postSavingLock reducer

adamsilverstein commented:
> We should have unit tests for at least selectors and reducer.

@gziolo Added tests. 

@aduth addressed your feedback, can you please give this another review?

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -1995,6 +1995,17 @@ export function isPostLocked( state ) {
+ * @return {boolean} Is locked.
+ */
+export function isPostSavingLocked( state ) {
+	return find( state.postSavingLock, ( lock ) => lock ) ? true : false;
```
My hunch upon seeing this is: Why can't this just be a condition of `Object.keys( state.postSavingLock ).length > 0` ? Why do we hang on to locks in state after the unlock occurs?

Which would require further revision to the reducer to use [`omit`](https://lodash.com/docs/4.17.10#omit):

```js
case 'LOCK_POST_SAVING':
	return { ...state, [ action.lockName ]: true };

case 'UNLOCK_POST_SAVING':
	return omit( state, action.lockName );
```

adamsilverstein replied:
yea, that bothered me as well, good suggest - i will refine that

adamsilverstein replied:
Adjusted in https://github.com/WordPress/gutenberg/pull/10649/commits/949ed50506190992c40c175891bc3a397cd6c3ba. Any reason to use `Object.keys( state.postSavingLock ).length > 0` instead of `state.postSavingLock.length > 0`?

CGlingener replied:
@adamsilverstein there is a very good reason: Object.length is undefined. 🙂 Opened a pull request to fix this little glitch https://github.com/WordPress/gutenberg/pull/11636

aduth commented on packages/editor/src/store/test/reducer.js:
```diff
@@ -2263,4 +2264,53 @@ describe( 'state', () => {
+				'test-lock': true,
+			} );
+
+			state = postSavingLock( state, {
```
Minor: It's generally a good practice in reducer tests to apply `deepFreeze` to the `state` passed to the reducer. It will indirectly protect against unintentional mutations (via thrown errors failing the tests). As written here, technically this test would be satisfied by the following theoretical misbehaving reducer:

```js
export function postSavingLock( state = {}, action ) {
	switch ( action.type ) {
		case 'LOCK_POST_SAVING':
		case 'UNLOCK_POST_SAVING':
			state[ action.lockName ] = action.type === 'LOCK_POST_SAVING';
			break;

	return state;
}
```

adamsilverstein replied:
added in https://github.com/WordPress/gutenberg/pull/10649/commits/0696c0c9288ba5b8c9994a6c3b09be79abc4d65f

aduth commented on packages/editor/src/store/reducer.js:
```diff
@@ -913,6 +913,29 @@ export function postLock( state = { isLocked: false }, action ) {
+ * @return {PostLockState} Updated state.
+ */
+export function postSavingLock( state = {}, action ) {
+	const { lockName } = action;
```
Minor: Since a reducer is called on every single action, it's generally best to avoid doing anything until you're certain to be working with a relevant action. On its own, the allocation here is not so concerning, but an unfamiliar future maintainer may perceive this to imply that logic targeting the lock-specific handling is okay to occur prior to switch case, and feel comfortable to make enhancements that at best harm performance and at worst introduce bugs.

tl;dr Best to move into the `case` handler(s).

adamsilverstein replied:
Adjusted in https://github.com/WordPress/gutenberg/pull/10649/commits/50d5a49c906e9c724f0fc06644fdf323a9f1ace1

aduth reviewed (approved):
My remaining feedback would be good to consider, but are non-blocking improvements.

adamsilverstein committed:
postSavingLock: deepFreeze( state ) passed in tests

adamsilverstein committed:
simplify reducers, selectors & adjust tests

adamsilverstein committed:
move lockName into case handlers

adamsilverstein commented:
@aduth thanks for the detailed feedback, I think I have addressed your suggestions, one outstanding question then we can merge: https://github.com/WordPress/gutenberg/pull/10649#discussion_r228360860

gziolo commented:
Thanks @adamsilverstein, great to see we have it finally sorted out 🎉

chrisvanpatten commented:
@adamsilverstein We've had a few requests for documentation on this so I'm digging in to try to understand how this was implemented. One question I have is why was this implemented just as a save lock, vs separate publish and save locks? Just curious from a conceptual level.

jaboi commented:
@chrisvanpatten Any update on documentation for this? I have a theme that requires a featured image to be horizontal, and the hooks I used to handle the validation and error messaging for this do not work at all with Gutenberg. I think this "save locking" feature is what I'm looking for, but need some kind of documentation to implement. Any feedback would be helpful. Thanks!

jaboi commented:
Does anyone have a working example of this save lock control action for a publish/update conditional?

adamsilverstein commented:
> We've had a few requests for documentation on this so I'm digging in to try to understand how this was implemented. One question I have is why was this implemented just as a save lock, vs separate publish and save locks? Just curious from a conceptual level. I have read the original ticket and the other related closed PRs, but still not sure I understand.

@chrisvanpatten missed this earlier. likely we chose this approach because this is how the buttons were already architected in regards to other conditionals. plugins should be able to control each button (publish vs update) individually still, happy to help with some example code here.

bryanbartow commented:
Any update on documentation and/or example(s) for this? Can't seem to find anything here, on the codex, nor stack exchange.

adamsilverstein commented:
> Any update on documentation and/or example(s) for this? Can't seem to find anything here, on the codex, nor stack exchange.

Happy to add some documentation, where should this go?

aduth commented:
> Happy to add some documentation, where should this go?

Since this was merged, enhancements were made to the data documentation workflow, so we could include examples for actions (like the ones introduced here) using `@example` JSDoc tags.

See also:

- #15421
- [Example `@example`](https://github.com/WordPress/gutenberg/blob/64a8a7c5208324b5308d2298bf0f98c906cfd66f/packages/a11y/src/index.js#L41-L50)

adamsilverstein commented:
> so we could include examples for actions (like the ones introduced here) using @example JSDoc tags.

Excellent, thanks for the tip. I'll work on it.
